plugin/trino-kafka-event-listener/src/main/java/io/trino/plugin/eventlistener/kafka/KafkaRecordBuilder.java

```diff
@@ -0,0 +1,102 @@
+import com.fasterxml.jackson.datatype.jdk8.Jdk8Module;
+import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;
+import io.trino.plugin.eventlistener.kafka.metadata.MetadataProvider;
+import io.trino.plugin.eventlistener.kafka.model.QueryCompletedEventWrapper;
+import io.trino.plugin.eventlistener.kafka.model.QueryCreatedEventWrapper;
+import io.trino.spi.eventlistener.QueryCompletedEvent;
+import io.trino.spi.eventlistener.QueryCreatedEvent;
+import org.apache.avro.Schema;
+import org.apache.avro.generic.GenericData;
+import org.apache.avro.generic.GenericRecord;
+import org.apache.avro.reflect.ReflectData;
+import org.apache.kafka.clients.producer.ProducerRecord;
+import org.apache.kafka.common.serialization.Serializer;
+
+import java.nio.charset.StandardCharsets;
+import java.util.Set;
+
+public class KafkaRecordBuilder
+        implements Serializer<Object>
+{
+    private static final Schema QUERY_COMPLETED_EVENT_SCHEMA =
+            ReflectData.get().getSchema(QueryCompletedEventWrapper.class);
+    private static final Schema QUERY_STARTED_EVENT_SCHEMA =
+            ReflectData.get().getSchema(QueryCreatedEventWrapper.class);
+
+    private final ObjectWriter writer;
+    private final String startedTopic;
+    private final String completedTopic;
+    private final MetadataProvider metadataProvider;
+
+    @JsonFilter("filter properties by name")
+    static class PropertyFilterMixIn {}
+
+    KafkaRecordBuilder(String startedTopic, String completedTopic, Set<String> excludedFields, MetadataProvider metadataProvider)
+    {
+        this.startedTopic = startedTopic;
+        this.completedTopic = completedTopic;
+        FilterProvider filter = new SimpleFilterProvider().addFilter("filter properties by name",
+                SimpleBeanPropertyFilter.serializeAllExcept(excludedFields));
+        this.writer = new ObjectMapper()
+                .registerModule(new Jdk8Module())
+                .registerModule(new JavaTimeModule())
+                .addMixIn(Object.class, PropertyFilterMixIn.class)
+                .writer(filter);
+        this.metadataProvider = metadataProvider;
+    }
+
+    @Override
+    public byte[] serialize(String topic, Object data)
+    {
+        return writeJson(data).getBytes(StandardCharsets.UTF_8);
+    }
+
+    public ProducerRecord<String, GenericRecord> buildStartedRecord(QueryCreatedEvent event)
+    {
+        GenericRecord eventRecord = new GenericData.Record(QUERY_STARTED_EVENT_SCHEMA);
+        eventRecord.put(0, writeJson(metadataProvider.getMetadata()));
+        eventRecord.put(1, writeJson(event));
+        return new ProducerRecord<>(startedTopic, eventRecord);
+    }
+
+    public ProducerRecord<String, GenericRecord> buildCompletedRecord(QueryCompletedEvent event)
+    {
+        GenericRecord eventRecord = new GenericData.Record(QUERY_COMPLETED_EVENT_SCHEMA);
+        eventRecord.put(0, writeJson(metadataProvider.getMetadata()));
+        eventRecord.put(1, writeJson(event));
```

Comment:
This is potentially problematic. By default Kafka message cannot be larger than 1MB and for bigger queries serialized JSON could exceed that. There are some fields which are verbose and large by nature:
* `QueryMetadata#query` -> some queries can be huge
* `QueryMetadata#plan`, `QueryMetadata#jsonPlan`. 
* `QueryStatistics#operatorSummariesProvider` contains `TDigestHistogram#digest` in `OperatorStats#metrics` (and other) fields.

You can increase Kafka message size, but this is not generally recommended. I think there are couple of options:
* ignore too big message errors (most likely not OK)
* trim fields in case of too big message (we loose some info for some queries)
* Fetch big fields query `QueryInfo` since `QueryMetadata` contains backreference: `QueryMetadata#uri`. You can reconstruct plan like that either.

I would guess the issues with big Kafka messages will only appear at scale and under bigger load.